File: Assets/Warlock/Scripts/Mvp/MvpRuntime.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Warlock.Mvp
{
    public readonly struct RoundUiState
    {
        public RoundUiState(RoundState phase, string title, string primaryAction, string roundText)
        {
            Phase = phase;
            Title = title;
            PrimaryAction = primaryAction;
            RoundText = roundText;
        }

        public RoundState Phase { get; }
        public string Title { get; }
        public string PrimaryAction { get; }
        public string RoundText { get; }
    }

    public static class RoundTransitionPresenter
    {
        public static RoundUiState Present(RoundState phase, int currentRound, int roundsTotal)
        {
            var roundText = currentRound > 0 && roundsTotal > 0
                ? $"Round {currentRound}/{roundsTotal}"
                : "Round N/A";

            return phase switch
            {
                RoundState.Waiting => new RoundUiState(phase, "Waiting For Players", "HostStartMatchWhenReady", roundText),
                RoundState.RoundStart => new RoundUiState(phase, "Round Start", "RoundIntroCountdown", roundText),
                RoundState.Combat => new RoundUiState(phase, "Combat", "EnableSkillCasting", roundText),
                RoundState.RoundEnd => new RoundUiState(phase, "Round End", "ShowRoundRanking", roundText),
                RoundState.Shop => new RoundUiState(phase, "Shop", "PurchaseAndEquipSkills", roundText),
                RoundState.MatchEnd => new RoundUiState(phase, "Match End", "ShowFinalWinner", roundText),
                _ => throw new ArgumentOutOfRangeException(nameof(phase), $"Unknown phase: {phase}")
            };
        }
    }

    public readonly struct BindingValidationResult
    {
        public BindingValidationResult(bool ok, IReadOnlyList<string> errors)
        {
            Ok = ok;
            Errors = errors;
        }

        public bool Ok { get; }
        public IReadOnlyList<string> Errors { get; }
    }

    public static class InputBindingRules
    {
        public static Dictionary<string, string> CreateDefaultSlotBindings()
        {
            return MvpConstants.LoadoutSlotKeys.ToDictionary(slot => slot, slot => slot);
        }

        public static BindingValidationResult Validate(IReadOnlyDictionary<string, string> slotToKey)
        {
            var errors = new List<string>();

            foreach (var slot in MvpConstants.LoadoutSlotKeys)
            {
                if (!slotToKey.ContainsKey(slot))
                {
                    errors.Add($"MISSING_SLOT_{slot}");
                }
            }

            if (slotToKey.Count != MvpConstants.LoadoutSlotKeys.Count)
            {
                errors.Add("SLOT_COUNT_MISMATCH");
            }

            var normalizedKeys = MvpConstants.LoadoutSlotKeys
                .Select(slot => slotToKey.TryGetValue(slot, out var key) ? key?.Trim().ToUpperInvariant() ?? string.Empty : string.Empty)
                .ToList();

            if (normalizedKeys.Any(string.IsNullOrWhiteSpace))
            {
                errors.Add("INVALID_KEY");
            }

            if (normalizedKeys.Distinct(StringComparer.Ordinal).Count() != normalizedKeys.Count)
            {
                errors.Add("DUPLICATE_KEY_BINDING");
            }

            return new BindingValidationResult(errors.Count == 0, errors);
        }

        public static (bool ok, string? reason, Dictionary<string, string>? bindings) Rebind(
            IReadOnlyDictionary<string, string> slotToKey,
            string slot,
            string key)
        {
            if (!MvpConstants.LoadoutSlotKeys.Contains(slot))
            {
                return (false, "INVALID_SLOT", null);
            }

            if (string.IsNullOrWhiteSpace(key))
            {
                return (false, "INVALID_KEY", null);
            }

            var next = slotToKey.ToDictionary(entry => entry.Key, entry => entry.Value);
            next[slot] = key.Trim().ToUpperInvariant();

            var validation = Validate(next);
            if (!validation.Ok)
            {
                return (false, validation.Errors[0], null);
            }

            return (true, null, next);
        }
    }

    public sealed class LoadoutUiModel
    {
        private readonly ISet<string> _ownedSkills;
        private readonly Dictionary<string, string?> _loadout;

        public LoadoutUiModel(ISet<string> ownedSkills, Dictionary<string, string?> loadout)
        {
            _ownedSkills = ownedSkills;
            _loadout = loadout;
        }

        public RuleResult Equip(string slot, string skillId)
        {
            return LoadoutRules.EquipSkill(_ownedSkills, _loadout, slot, skillId);
        }

        public RuleResult Unequip(string slot)
        {
            if (!MvpConstants.LoadoutSlotKeys.Contains(slot))
            {
                return RuleResult.Fail("INVALID_SLOT");
            }

            _loadout[slot] = null;
            return RuleResult.Success();
        }

        public IReadOnlyDictionary<string, string?> Snapshot()
        {
            return new Dictionary<string, string?>(_loadout);
        }
    }

    public readonly struct HostRelayEvent
    {
        public HostRelayEvent(int seq, string type, IReadOnlyDictionary<string, object> payload)
        {
            Seq = seq;
            Type = type;
            Payload = payload;
        }

        public int Seq { get; }
        public string Type { get; }
        public IReadOnlyDictionary<string, object> Payload { get; }
    }

    public sealed class HostRelaySession
    {
        private static int _roomSequence;
        private static readonly HashSet<string> KnownEvents = new(
            NetworkContract.MatchEvents
                .Concat(NetworkContract.RoundEvents)
                .Concat(NetworkContract.CombatEvents)
                .Concat(NetworkContract.EconomyEvents)
                .Concat(NetworkContract.ArenaEvents),
            StringComparer.Ordinal
        );

        private readonly List<HostRelayEvent> _events = new();
        private readonly HashSet<string> _players = new(StringComparer.Ordinal);
        private readonly HashSet<string> _disconnectedPlayers = new(StringComparer.Ordinal);
        private int _lastSeq;

        public HostRelaySession(string hostId)
        {
            HostId = string.IsNullOrWhiteSpace(hostId)
                ? throw new ArgumentException("hostId is required", nameof(hostId))
                : hostId;
        }

        public string HostId { get; }
        public string? RoomId { get; private set; }
        public string? InviteCode { get; private set; }
        public RoomSettings? Settings { get; private set; }

        public (bool ok, string? reason) CreateRoom(int? maxPlayers, int? rounds)
        {
            var validation = RoomSettingsValidator.Validate(maxPlayers, rounds);
            if (!validation.Ok)
            {
                return (false, validation.Errors.FirstOrDefault() ?? "INVALID_ROOM_SETTINGS");
            }

            _roomSequence += 1;
            RoomId = $"room-{_roomSequence:0000}";
            InviteCode = $"WLK{_roomSequence:0000}";
            Settings = validation.Settings;
            _players.Clear();
            _disconnectedPlayers.Clear();
            _players.Add(HostId);

            PushEvent("RoomCreated", new Dictionary<string, object>
            {
                { "roomId", RoomId },
                { "inviteCode", InviteCode },
                { "hostId", HostId }
            });

            return (true, null);
        }

        public (bool ok, string? reason) JoinPlayer(string playerId)
        {
            if (RoomId == null || Settings == null)
            {
                return (false, "ROOM_NOT_CREATED");
            }

            if (string.IsNullOrWhiteSpace(playerId))
            {
                return (false, "INVALID_PLAYER_ID");
            }

            if (_players.Contains(playerId))
            {
                return (true, null);
            }

            if (_players.Count >= Settings.Value.MaxPlayers)
            {
                return (false, "ROOM_FULL");
            }

            _players.Add(playerId);
            PushEvent("PlayerJoined", new Dictionary<string, object>
            {
                { "roomId", RoomId },
                { "playerId", playerId }
            });
            return (true, null);
        }

        public (bool ok, string? reason, HostRelayEvent? @event) SyncEvent(string eventType, IReadOnlyDictionary<string, object>? payload = null)
        {
            if (!KnownEvents.Contains(eventType))
            {
                return (false, "UNKNOWN_EVENT_TYPE", null);
            }

            var created = PushEvent(eventType, payload ?? new Dictionary<string, object>());
            return (true, null, created);
        }

        public IReadOnlyList<HostRelayEvent> ReadEventsSince(int seq)
        {
            return _events.Where(evt => evt.Seq > seq).ToList();
        }

        public (bool ok, string action) OnDisconnect(string playerId)
        {
            if (!_players.Contains(playerId))
            {
                return (false, "REJECT_UNKNOWN_PLAYER");
            }

            if (playerId == HostId)
            {
                PushEvent("MatchEnded", new Dictionary<string, object>
                {
                    { "reason", "HOST_DISCONNECTED" }
                });
                return (true, "SAFE_TERMINATE");
            }

            _disconnectedPlayers.Add(playerId);
            PushEvent("PlayerEliminated", new Dictionary<string, object>
            {
                { "playerId", playerId },
                { "reason", "PLAYER_DISCONNECTED" }
            });
            return (true, "MARK_INACTIVE");
        }

        public (bool ok, string action, int resumeFromSeq) OnReconnect(string playerId)
        {
            if (!_players.Contains(playerId))
            {
                return (false, "REJECT_UNKNOWN_PLAYER", 0);
            }

            if (!_disconnectedPlayers.Contains(playerId))
            {
                return (true, "NO_OP", _lastSeq);
            }

            _disconnectedPlayers.Remove(playerId);
            PushEvent("PlayerJoined", new Dictionary<string, object>
            {
                { "playerId", playerId },
                { "reconnected", true },
                { "resumeFromSeq", _lastSeq }
            });
            return (true, "RESUME_FROM_LAST_SEQUENCE", _lastSeq);
        }

        private HostRelayEvent PushEvent(string type, IReadOnlyDictionary<string, object> payload)
        {
            _lastSeq += 1;
            var created = new HostRelayEvent(_lastSeq, type, payload);
            _events.Add(created);
            return created;
        }
    }
}

File: src/mvp-runtime.js
import { LOADOUT_SLOT_KEYS, ROUND_STATES } from './constants.js';
import { equipSkill, unequipSkill } from './loadout.js';
import { NETWORK_EVENTS } from './network-contract.js';
import { validateRoomSettings } from './room-settings.js';

const ROUND_UI_DEFINITION = Object.freeze({
  [ROUND_STATES.WAITING]: Object.freeze({
    title: 'Waiting For Players',
    primaryAction: 'HostStartMatchWhenReady',
  }),
  [ROUND_STATES.ROUND_START]: Object.freeze({
    title: 'Round Start',
    primaryAction: 'RoundIntroCountdown',
  }),
  [ROUND_STATES.COMBAT]: Object.freeze({
    title: 'Combat',
    primaryAction: 'EnableSkillCasting',
  }),
  [ROUND_STATES.ROUND_END]: Object.freeze({
    title: 'Round End',
    primaryAction: 'ShowRoundRanking',
  }),
  [ROUND_STATES.SHOP]: Object.freeze({
    title: 'Shop',
    primaryAction: 'PurchaseAndEquipSkills',
  }),
  [ROUND_STATES.MATCH_END]: Object.freeze({
    title: 'Match End',
    primaryAction: 'ShowFinalWinner',
  }),
});

function assertKnownRoundState(phase) {
  if (!ROUND_UI_DEFINITION[phase]) {
    throw new Error(`Unknown round phase: ${phase}`);
  }
}

export function presentRoundTransition({ phase, currentRound = 0, roundsTotal = 0 } = {}) {
  assertKnownRoundState(phase);
  const base = ROUND_UI_DEFINITION[phase];
  const roundText =
    roundsTotal > 0 && currentRound > 0 ? `Round ${currentRound}/${roundsTotal}` : 'Round N/A';
  return {
    phase,
    title: base.title,
    primaryAction: base.primaryAction,
    roundText,
  };
}

export function createDefaultSlotBindings() {
  return Object.fromEntries(LOADOUT_SLOT_KEYS.map((slot) => [slot, slot]));
}

export function validateSlotBindings(slotToKey) {
  const errors = [];
  const slots = Object.keys(slotToKey ?? {});

  for (const slot of LOADOUT_SLOT_KEYS) {
    if (!slots.includes(slot)) {
      errors.push(`MISSING_SLOT_${slot}`);
    }
  }

  if (slots.length !== LOADOUT_SLOT_KEYS.length) {
    errors.push('SLOT_COUNT_MISMATCH');
  }

  const keys = LOADOUT_SLOT_KEYS.map((slot) => slotToKey?.[slot]);
  const normalizedKeys = keys.map((key) => (typeof key === 'string' ? key.trim().toUpperCase() : ''));

  if (normalizedKeys.some((key) => key.length === 0)) {
    errors.push('INVALID_KEY');
  }

  if (new Set(normalizedKeys).size !== normalizedKeys.length) {
    errors.push('DUPLICATE_KEY_BINDING');
  }

  return errors.length === 0
    ? { ok: true, bindings: Object.freeze({ ...slotToKey }) }
    : { ok: false, errors };
}

export function rebindSlot(slotToKey, { slot, key }) {
  if (!LOADOUT_SLOT_KEYS.includes(slot)) {
    return { ok: false, reason: 'INVALID_SLOT' };
  }

  if (typeof key !== 'string' || key.trim().length === 0) {
    return { ok: false, reason: 'INVALID_KEY' };
  }

  const next = { ...slotToKey, [slot]: key.trim().toUpperCase() };
  const validation = validateSlotBindings(next);
  if (!validation.ok) {
    return { ok: false, reason: validation.errors[0], errors: validation.errors };
  }

  return { ok: true, bindings: next };
}

export class LoadoutUiModel {
  constructor({ ownedSkills, loadout }) {
    this.ownedSkills = ownedSkills;
    this.loadout = loadout;
  }

  equip(slot, skillId) {
    return equipSkill({
      ownedSkills: this.ownedSkills,
      loadout: this.loadout,
      slot,
      skillId,
    });
  }

  unequip(slot) {
    return unequipSkill({ loadout: this.loadout, slot });
  }

  snapshot() {
    return Object.freeze({ ...this.loadout });
  }
}

const ALL_NETWORK_EVENT_TYPES = new Set([
  ...NETWORK_EVENTS.match,
  ...NETWORK_EVENTS.round,
  ...NETWORK_EVENTS.combat,
  ...NETWORK_EVENTS.economy,
  ...NETWORK_EVENTS.arena,
]);

let roomSequence = 0;

function nextRoomId() {
  roomSequence += 1;
  return `room-${String(roomSequence).padStart(4, '0')}`;
}

function inviteCodeFromRoomId(roomId) {
  const suffix = roomId.replace('room-', '');
  return `WLK${suffix}`;
}

export class HostRelaySession {
  constructor({ hostId }) {
    if (!hostId) {
      throw new Error('hostId is required');
    }

    this.hostId = hostId;
    this.room = null;
    this.players = new Set();
    this.disconnectedPlayers = new Set();
    this.events = [];
    this.lastSeq = 0;
  }

  createRoom({ maxPlayers, rounds } = {}) {
    const validation = validateRoomSettings({ maxPlayers, rounds });
    if (!validation.ok) {
      return { ok: false, reason: validation.errors[0], errors: validation.errors };
    }

    const roomId = nextRoomId();
    this.room = {
      id: roomId,
      inviteCode: inviteCodeFromRoomId(roomId),
      settings: validation.settings,
    };
    this.players = new Set([this.hostId]);
    this.disconnectedPlayers.clear();

    this._pushEvent('RoomCreated', {
      roomId: this.room.id,
      inviteCode: this.room.inviteCode,
      hostId: this.hostId,
    });

    return { ok: true, room: this.room };
  }

  joinPlayer(playerId) {
    if (!this.room) {
      return { ok: false, reason: 'ROOM_NOT_CREATED' };
    }

    if (!playerId) {
      return { ok: false, reason: 'INVALID_PLAYER_ID' };
    }

    if (this.players.has(playerId)) {
      return { ok: true, alreadyJoined: true };
    }

    if (this.players.size >= this.room.settings.maxPlayers) {
      return { ok: false, reason: 'ROOM_FULL' };
    }

    this.players.add(playerId);
    this._pushEvent('PlayerJoined', { roomId: this.room.id, playerId });
    return { ok: true };
  }

  syncEvent(eventType, payload = {}) {
    if (!ALL_NETWORK_EVENT_TYPES.has(eventType)) {
      return { ok: false, reason: 'UNKNOWN_EVENT_TYPE' };
    }

    const event = this._pushEvent(eventType, payload);
    return { ok: true, event };
  }

  readEventsSince(seq = 0) {
    return this.events.filter((event) => event.seq > seq);
  }

  onDisconnect(playerId) {
    if (!this.players.has(playerId)) {
      return { ok: false, action: 'REJECT_UNKNOWN_PLAYER' };
    }

    if (playerId === this.hostId) {
      this._pushEvent('MatchEnded', {
        reason: 'HOST_DISCONNECTED',
      });
      return {
        ok: true,
        action: 'SAFE_TERMINATE',
      };
    }

    this.disconnectedPlayers.add(playerId);
    this._pushEvent('PlayerEliminated', {
      playerId,
      reason: 'PLAYER_DISCONNECTED',
    });
    return {
      ok: true,
      action: 'MARK_INACTIVE',
    };
  }

  onReconnect(playerId) {
    if (!this.players.has(playerId)) {
      return { ok: false, action: 'REJECT_UNKNOWN_PLAYER' };
    }

    if (!this.disconnectedPlayers.has(playerId)) {
      return { ok: true, action: 'NO_OP' };
    }

    this.disconnectedPlayers.delete(playerId);
    this._pushEvent('PlayerJoined', {
      playerId,
      reconnected: true,
      resumeFromSeq: this.lastSeq,
    });
    return {
      ok: true,
      action: 'RESUME_FROM_LAST_SEQUENCE',
      resumeFromSeq: this.lastSeq,
    };
  }

  _pushEvent(type, payload) {
    this.lastSeq += 1;
    const event = {
      seq: this.lastSeq,
      type,
      payload,
    };
    this.events.push(event);
    return event;
  }
}
